stackslib/src/chainstate/nakamoto/miner.rs

```diff
@@ -833,3 +835,252 @@ impl BlockBuilder for NakamotoBlockBuilder {
         result
     }
 }
+
+/// This enum is used to supply a `reason_code` for validation
+///  rejection responses. This is serialized as an enum with string
+///  type (in jsonschema terminology).
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum ValidateRejectCode {
+    BadBlockHash,
+    BadTransaction,
+    InvalidBlock,
+    ChainstateError,
+    UnknownParent,
+}
+
+/// A response for block proposal validation
+///  that the stacks-node thinks should be rejected.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateReject {
+    pub reason: String,
+    pub reason_code: ValidateRejectCode,
+}
+
+/// A response for block proposal validation
+///  that the stacks-node thinks is acceptable.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateOk {
+    pub block: NakamotoBlock,
+    pub cost: ExecutionCost,
+    pub size: u64,
+}
+
+/// This enum is used for serializing the response to block
+/// proposal validation.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "Result")]
+pub enum BlockValidateResponse {
+    Ok(BlockValidateOk),
+    Reject(BlockValidateReject),
+}
+
+impl From<Result<BlockValidateOk, BlockValidateReject>> for BlockValidateResponse {
+    fn from(value: Result<BlockValidateOk, BlockValidateReject>) -> Self {
+        match value {
+            Ok(o) => BlockValidateResponse::Ok(o),
+            Err(e) => BlockValidateResponse::Reject(e),
+        }
+    }
+}
+
+impl From<Error> for BlockValidateReject {
+    fn from(value: Error) -> Self {
+        BlockValidateReject {
+            reason: format!("Chainstate Error: {value}"),
+            reason_code: ValidateRejectCode::ChainstateError,
+        }
+    }
+}
+
+impl From<DBError> for BlockValidateReject {
+    fn from(value: DBError) -> Self {
+        Error::from(value).into()
+    }
+}
+
+/// Represents a block proposed to the `v2/block_proposal` endpoint for validation
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct NakamotoBlockProposal {
+    /// Proposed block
+    pub block: NakamotoBlock,
+    // tenure ID -- this is the index block hash of the start block of the last tenure (i.e.
+    // the data we committed to in the block-commit).  If this is an epoch 2.x parent, then
+    // this is just the index block hash of the parent Stacks block.
+    pub tenure_start_block: StacksBlockId,
+    /// Identifies which chain block is for (Mainnet, Testnet, etc.)
+    pub chain_id: u32,
+}
+
+impl StacksMessageCodec for NakamotoBlockProposal {
```

Comment:
I'd recommend just deleting the `StacksMessageCodec` implementation here. Having two representations of the serialization for the proposal seems unnecessary, and JSON via serde is preferable.

Reply:
I'm not sure I even need the `NakamotoBlockProposal` struct anyway. I've already removed most of the fields from the original `BlockProposal` struct in Subnets. I suppose I could just look up the most recent `tenure_start_block`, and a block proposal for the wrong `chain_id` would be invalid anyways. Would be simpler for miners if the endpoint just takes `NakamotoBlock`. What do you think?

Reply:
I think it's best to keep a `NakamotoBlockProposal` struct even if it just contains `block` for a couple reasons:

1. The JSON interface will make it less painful in the future if we decide there's some other information that's important to supply to the proposal validation.
2. Having the `validate()` method be a member of a new proposal struct avoids continuing the expansion of some already fairly unwieldy interfaces.

Reply:
I'm removing `tenure_start_block` because that isn't used by `NakamotBlockBuilder` anymore. So it'll just have `block` and `chain_id` for now

> The JSON interface will make it less painful in the future if we decide there's some other information that's important to supply to the proposal validation

I'm having a hard time thinking of anything that might be useful for block validation but not already contained in the block. That's the whole point of a blockchain, isn't it, to be able to validate the entire history from just the blocks?